### PR TITLE
Overhaul linting and formatting

### PR DESCRIPTION
**Pull request by niksirbi**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

We use `ruff` for linting and `black` for auto-formatting. With the introduction of `ruff.format` (a drop-in replacement for `black`, but faster) there is no longer need to be using `black`. This prompted me to overhaul our entire lint/format setup.

**What does this PR do?**

I went through the [scientific python development guide for style and static checks](https://learn.scientific-python.org/development/guides/style/) and applied the recommendations that make sense for `movement`.

In summary these are:
- switch from `black` to `ruff.format`
- add some more [ruff rules](https://docs.astral.sh/ruff/rules/). I ended up with the following set (which includes the most popular ruff rules):
	```yaml
	lint.select = [
	  "E",   # pycodestyle errors
	  "F",   # Pyflakes
	  "UP",  # pyupgrade
	  "I",   # isort
	  "B",   # flake8 bugbear
	  "SIM", # flake8 simplify
	  "C90", # McCabe complexity
	]
    ```
    Of those, we were already using `E`, `F`, and `I`. With the addition of `UP`, `B`, `SIM`, and `C90` I discovered and fixed a few minor problems that the other checks had missed. That said, there were only a few extra issues, which means we are already great Python coders 😉 .
 - add some more recommended pre-commit hooks we were missing, such as `check-added-large-files` and `check-yaml`.
 - add pre-commit hooks to specifically catch some easy-to-make rst mistakes (e.g. single-vs-double backticks for monospace formatting)

**What does this PR NOT do?**
I also considered adding the "D" `ruff` rule (pydocstyle) to enforce consistent formatting for our docstrings (according to PEP8 recommendations). I think this is super useful, but currently requires some manual work to update lots of our non-compliant docstrings (135 non auto-fixable violations!). I will open an issue to do this later.

## References

If we end up being happy with the new setup, we may consider [porting it over to the python-cookiecutter](https://github.com/neuroinformatics-unit/python-cookiecutter/issues/73).

## Does this PR require an update to the documentation?

The contributed guidelines have been updated accordingly.

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by niksirbi** [comment]
Tagging @neuroinformatics-unit/behaviour and @b-peri, because these changes will affect workflows for all.

File: pyproject.toml [hunk 1]
```diff
@@ -100,9 +94,20 @@ ignore_missing_imports = true
 [tool.ruff]
 line-length = 79
 exclude = ["__init__.py", "build", ".eggs"]
-lint.select = ["I", "E", "F"]
+lint.select = [
```
Comment (niksirbi) [inline pyproject.toml:97]: add "D" commented out for now

**Comment by lochhh** [comment]
Thanks again @niksirbi ! Happy to merge once the following two things are addressed: 
- [ ] The `mixed-line-ending` check fails with 52 files when running `pre-commit run --all-files` locally on Windows. Could you try this on a Mac? 
- [ ] Add the "D" ruff rule (pydocstyle) but commented out in prep for the PR to fix the violations.

**Review by lochhh — APPROVED** [review]

File: pyproject.toml
[the remark this answers is not in the document]
Comment (niksirbi) [reply pyproject.toml:1]: Thaank for the comment. But we are in fact already re-running the pre-commit as part of out GitHub actions workflow:
```yaml
jobs:
  linting:
    runs-on: ubuntu-latest
    steps:
      - uses: neuroinformatics-unit/actions/lint@v2
 ```
 
 That action takes care of re-running the linting during CI.

**Comment by adamltyson** [comment]
Assuming these changes aren't too specific to `movement`, could they be added to the cookiecutter, and gradually rolled out to other packages? I'm very keen that our tools don't start to diverge in terms of tooling.

**Comment by niksirbi** [comment]
> Assuming these changes aren't too specific to `movement`, could they be added to the cookiecutter, and gradually rolled out to other packages? I'm very keen that our tools don't start to diverge in terms of tooling.

Yes, the plan was to roll them out in `movement` first, for testing, and after a few weeks (assuming all is well) roll them out to the cookiecutter + other packages.

**Comment by niksirbi** [comment]
> * The `mixed-line-ending` check fails with 52 files when running `pre-commit run --all-files` locally on Windows. Could you try this on a Mac?

@lochhh I can't reproduce this on Mac. Running `pre-commit run --all-files` on my M2 Mac finds no issues. Could you check whether your IDE is configured to automatically use CRLF for line endings (i.e. `\r\n`,  that's the Windows way)? On Unix systems it's LF (`\n`) and that's what the pre-commit hook enforces (and in theory, auto-fixes).

**Comment by lochhh** [comment]
> > * The `mixed-line-ending` check fails with 52 files when running `pre-commit run --all-files` locally on Windows. Could you try this on a Mac?
> 
> @lochhh I can't reproduce this on Mac. Running `pre-commit run --all-files` on my M2 Mac finds no issues. Could you check whether your IDE is configured to automatically use CRLF for line endings (i.e. `\r\n`, that's the Windows way)? On Unix systems it's LF (`\n`) and that's what the pre-commit hook enforces (and in theory, auto-fixes).

Thanks for looking into this. Can confirm it is indeed my IDE that's automatically converting the line endings.

**Comment by niksirbi** [comment]
Thanks for verifying @lochhh. I'll merge this now.

**Pull request merged**